### ISSUE TITLE
chain of never resolved promises create memory leaks

### ISSUE THREAD

**Issue by jonathanong** [promises-aplus/promises-spec#179]
is this the right place to ask this?

https://github.com/tj/co/issues/180#issuecomment-68094905

**Comment by Arnavion** [comment]
So is all that's left here to reword 2.3.2 to clarify how an own-promise is to be adopted? (And add the test in https://github.com/promises-aplus/promises-spec/issues/179#issuecomment-68211222 to the suite perhaps.)

**Comment by ForbesLindesay** [comment]
@Arnavion yes.  I believe you are correct.

**Comment by rkatic** [comment]
Interesting. I'm also considering to implement a similar "proxy" fix (I could be wrong, but if I recall correctly, Q v2 was doing that for some time already...)
I would agree with @briancavalier that the order should be undefined.

**Comment by petkaantonov** [comment]
I disagree with undefined order, 1,2,3, or the accidental order from straightforward proxy implementation (2,1,3). The order should be 2,3,1 because that's what it will be when `b` is a thenable or subclass.

**Comment by stefanpenner** [comment]
> The order should be 2,3,1 because that's what it will be when b is a thenable or subclass.

:100:

**Comment by rkatic** [comment]
So let me see if I'm understanding you. With A->B->C; C.then(f); B.then(g), resolveA();
your "proxy" fix will still call g before f?

**Comment by petkaantonov** [comment]
F is called before G in both cases where A is a promise or thenable.

**Comment by bergus** [comment]
I don't see any reason to specify an ordering here. The callbacks on `a` have nothing to do with those on `b`, so why should there any call sequence be required?
The only guarantees that the spec gives (and will ever need to give) is that in both `p.then(f).then(g)` and `p.then(f); p.then(g)` the `f` function is called before `g`.

**Comment by hax** [comment]
@rkatic I think order can be preserved, I will try to write another promise implementation to prove it :)

**Comment by domenic** [comment]
Can someone help me understand the action items from this thread a bit better? In particular I am interested in:

- What should change in the Promises/A+ spec
- What the observable differences are (there seem to be some ordering ones?). This would optimally be phrased in terms of a test case for the Promises/A+ test suite
- What should change in the ES6 promises spec. This would optimally be phrased in terms of a pull request for https://github.com/domenic/cancelable-promise/blob/master/lib/promise.js what makes the latter test pass.

**Comment by domenic** [comment]
Also, the title of this issue states it's about never-resolved promises. Is that still accurate? (Is it precise, terminology-wise?)

**Comment by ForbesLindesay** [comment]
I think the best way to think of this is as being a request for tail recursion on promises.  i.e.

```js
function run(i) {
  return i < 99999999 ? run(i + 1) : i;
}
console.log(run(0));
```

would run out of stack and cause an exception.  Similarly:

```js
function run(i) {
    return  new Promise(function(resolve) {
        setImmediate(resolve);
    }).then(function () {
      return i < 99999999 ? run(i + 1) : i;
    });
}
run(0).then(function (result) {
  console.log(result);
});
```

Maintains an ever increasing stack of promises, resulting in an out of memory error.

**Comment by ForbesLindesay** [comment]
One issue, in the case that `a` adopts the state of `b`, there are two options (and only two as far as I'm aware):

1. `a` moves all it's pending handlers to `b`, then `a` maintains a reference to `b`, all future calls to `a.then` attach their handlers directly to `b`.
2. `a` calls `b.then(fulfillA, rejectA)`, which causes `b` to maintain a reference to `a` and fulfill/reject it as appropriate later.  This is what the existing ES6 spec does.

The third option is to make use of weak-references in order to squash the chain in both directions.  It works something like this:


1. `a` moves all it's pending handlers to `b`
2. `a` maintains a reference to `b`
3. all future calls to `a.then` attach their handlers directly to `b`.
4. `b` maintains a weak reference to `a`
5. when `b` adopts the state of `c` it messages `a`, which replaces its reference to `b` with a reference to `c`.

At the end of that process, the head and tail (`a` and `c`) may be explicitly referenced by user code, without any code keeping a reference to `b`.

**Comment by ForbesLindesay** [comment]
As for test case, The only way I've figured out of testing that this is working is to profile the memory as in https://github.com/then/promise/blob/master/test/memory-leak.js

The ordering change is that for:

```js
var A = new Promise();
var B = new Promise();
RESOLVE(B, A);
B.then(function () {
  console.log('B');
});
A.then(function () {
  console.log('A');
});
RESOLVE(A, null);
```

would log B, A instead of A, B unless you do some clever workaround.

**Comment by bergus** [comment]
@domenic Let me try to put this down:

Nothing needs to change in the current Promises/A+ spec. There might be amendments to make, of varying strength, to the "adopt state" step, as point 2.3.2.4:
* "*When `x`is fulfilled, the `onFulfilled` callbacks of `promise` must be executed after the `onFulfilled` callbacks of `x`* that *were installed when the adoption happened.*"
* "*When `x`is fulfilled, the `onFulfilled` callbacks of `promise` must be executed after* all *the `onFulfilled` callbacks of `x` (at the time of the fulfillment).*"

(and respectively for rejection and `onRejected` callbacks, point 2.3.2.5).

The first specifies basically what would happen when a thenable would be assimilated, where a `then` call on `x` must execute the `resolvePromise`/`rejectPromise` callback that fulfills/rejects promise only after callbacks from previous `x.then` calls.
The second strengthens this to include further `x.then()` calls after the adoption. It basically describes the behaviour of thenable assimilation if we assume that `resolvePromise`/`rejectPromise` are asynchronously calling callbacks and callbacks are executed in a synchronous sequence (not given!). I would recommend to avoid this, as it prevents the "straightforward proxy implementation" (as Petka called it), and leaks memory for a deeply recursive adoption (`a` adopts `b` which adopts `c` which adopts `d`…), or at least makes implementation unnecessary complicated (haven't completed my thoughts on this yet).

We further might want to nail down the behaviour (callback ordering) when `x` adopted multiple times, by `promise1` and then by `promise2`. Should `promise1.then()` callbacks be executed before `promise2.then()` callbacks?

A test case could be derived from the examples given above (without the `"third"` callback).

Going to look at the ES6 spec now.

**Comment by bergus** [comment]
Ah thanks at @ForbesLindesay I obviously hadn't seen your post. Yes, that's exactly what we want to get. This third option with the "weak references" is just what I have implemented in [my own Promise library](https://github.com/bergus/F-Promise). And indeed, my library does do `B A`, and to change that it would require some heavy workaround.

**Comment by petkaantonov** [comment]
@bergus how do you do weak references? there is just a WeakMap. Are you using that?

**Comment by bergus** [comment]
@petkaantonov no, I use an "indirect reference". Promises that adopt each other share a common handle that contains a reference to innermost adopted promise (somewhere around https://github.com/bergus/F-Promise/blob/b66e917e2bf8836dab824875e77170f088e8361c/Promise.js#L151). This handle is `a`s only reference to `b`, and when `b` adopts `c` it changes the handle's `.resolution` to `c` without having a reference to `a`.

**Comment by bergus** [comment]
@domenic 

> What should change in the ES6 promises spec

I don't think anything specific needs to change. Except, if we wanted to spec tail-adoption-optimisation, probably everything would need to change (promise capabilities getting resolved lazily etc).

**Comment by domenic** [comment]
> I don't think anything specific needs to change. Except, if we wanted to spec tail-adoption-optimisation, probably everything would need to change (promise capabilities getting resolved lazily etc).

That basically gives me no new information. Do we need to spec tail-adoption-optimization to avoid the memory leak? If so what are the observable consequences---just ordering changes?

**Comment by bergus** [comment]
> Do we need to spec tail-adoption-optimization to avoid the memory leak?

Need? Probably not. I think the spec does already allow optimised implementations. Do we want to specify it? Maybe, I can't tell. That's a design decision, just as optimised tail recursion was.

> If so what are the observable consequences---just ordering changes?

Yes, if ordering was changed this could simplify a few things. I'm not sure whether it is required at all to avoid the memory leak, but I guess it would need some very clever implementation strategies to avoid the memory leak and guarantee the ordering specced by the PromiseJobs.

If this was changed, a possibly different ordering of callbacks would be the only observable difference (and your app not breaking, of course), the rest of the user-visible methods would behave the same.

**Comment by domenic** [comment]
OK, so my summary of where we are now with regard to ES6 is that people think it might be possible to maintain the current observable semantics with no spec changes and also avoid memory leaks, but they are unsure, and think that it would probably be easier to avoid memory leaks if we changed the ordering.

**Comment by ForbesLindesay** [comment]
That seems like a reasonable summary.  At the moment, A+ doesn't even specify this particular ordering.  I'm not sure about ES6 though.

I do think we should find a way to specify that this optimisation must be performed, because code that assumes it's implemented will break in really hard to debug ways if the implementation doesn't support it.

**Comment by bergus** [comment]
@domenic OK I reviewed this and found something that will need to change semantically (besides possibly the ordering)

    ∀x: Promise.resolve(x) ≡ Promise.resolve(Promise.resolve(x))

see https://mail.mozilla.org/pipermail/es-discuss/2015-April/042517.html

**Comment by domenic** [comment]
I think the following example is also symptomatic of the problem, without any recursion involved:

```js
const x0 = new Promise(r => global.resolveX0 = r);

const x1 = Promise.resolve(x0);
global.x2 = Promise.resolve(x1);
```

The reference chain here, even after GC, becomes `global.resolveX0 -> x0 -> x1 -> global.x2`. Wheres we would like `x1` to be GC-able, so that we have `global.resolveX0 -> x0 -> global.x2`.

This occurs because `global.x2 = Promise.resolve(x1)` translates to `x1.then(resolveX2, rejectX2)` (so `x1 -> global.x2`) and `const x1 = Promise.resolve(x0)` translates to `x0.then(resolveX1, rejectX1)` (so `x0 -> x1`).

Obviously you can add arbitrary intermediate levels to the chain.

**Comment by domenic** [comment]
@ForbesLindesay I do not understand (probably because I am missing something obvious) how your solution in https://github.com/domenic/cancelable-promise/pull/1 will work in the following case:

```js
const x0 = new Promise(r => global.resolveX0 = r);

const x1 = Promise.resolve(x0);
const x2 = Promise.resolve(x1);

x2.then(v => console.log(v));
```

If I read your solution correctly, the reference graph set up here is

```
global.resolveX0 -> x0
x1 -> x0
x2 -> x0
```

However after a GC run, all non-rooted chains will go away, resulting in

```
global.resolveX0 -> x0
```

i.e. the handler will never be executed.

But this must be wrong, because otherwise we have created a way of observing GC behavior in JavaScript, which is not otherwise possible. So I am missing something obvious. Please help :)

**Comment by domenic** [comment]
Oh, I see, I missed https://github.com/domenic/cancelable-promise/pull/1/files#diff-6230eb7fbf1dd2186bf1be6042c27f55R389, which essentially forwards the call `x2.then(v => console.log(v))` to `x0`. So the graph is really

```js
global.resolveX0 -> x0 -> (v => console.log(v))
x1 -> x0
x2 -> x0
```

**Comment by dr-js** [comment]
The issue title is a bit misleading, the tail-recursive code used for promise chain setup should be the problem.

I've done some test and the full comment with node test code is at: https://github.com/nodejs/node/issues/6673#issuecomment-599188223. And here's some quote:

> From what I tested, the pattern for promise memory leaking must include a tail-recursive setup, the promise creating function always recursively appear in itself's `promise.then` code.
> The recursive code cause the promise implementation need to keep extra reference/stack info so the GC can not claim those already-resolved promise from the chain.
> 
> If we build a long/infinite promise chain gradually without the recursive code pattern, and only keep reference to the tail of the chain, the GC can took those un-referenced promise from the chain, whether the promise is pending or not.
> 
> So I think chaining promise up is not always unsafe, it depends on the actual code.
> And if the promise version of `while (true) { /* async code */ }` is needed, better use a `async` function instead of the tail-recursive promise pattern.

**Comment by tsteinholz** [comment]
Hello,

I have been struggling with this issue myself, and the solution that I came to was the following:

```
const run = async () => {
    while (true) {
        await someAsyncFunction()
            .finally(() => { delete this; })  // Deallocate promise for GC.
    }
}

run();
```

I can't honestly say I understand the full ramifications for deleting this, but this change is the difference between allocating +1-7 kB/s in heap (creating a memory leak) and running the loop with +/- 1-2 kB/s in heap (alleviating the issue).
